lib/python/treadmill_aws/garbage_collector.py

```diff
@@ -0,0 +1,64 @@
+"""Garbage collector for LDAP and IPA.
+"""
+
+import logging
+from datetime import datetime
+from datetime import timedelta
+from socket import gaierror
+from socket import gethostbyname
+
+import jmespath
+
+from treadmill import admin
+from treadmill import context
+from treadmill_aws import awscontext
+from treadmill_aws.ipaclient import NotFoundError
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def servers():
+    """Remove terminated ec2 instances from LDAP and IPA"""
+    _LOGGER.info('Servers cleanup started')
+    admin_srv = admin.Server(context.GLOBAL.ldap.conn)
+    # One hour grace time for servers to bootstrap and register
+    one_hour_ago = datetime.now().utcnow() - timedelta(seconds=3600)
+    ldap_servers = set(
+        server[1].get('server').pop()
+        for server in admin_srv.admin.search(
+            search_base='ou=servers,%s' % admin_srv.admin.root_ou,
+            search_filter="(&(server=*)(createTimestamp<=%s))" %
+            one_hour_ago.strftime("%Y%m%d%H%M%SZ"),
+            attributes=["server"]))
+
+    _LOGGER.info('Fetching valid instances from AWS')
+    ec2 = jmespath.compile(
+        "Reservations[].Instances[].[Tags[?Key=='Name'].Value][][]")
+    ec2_instances = set(ec2.search(awscontext.GLOBAL.ec2.describe_instances(
+        Filters=[{"Name": "instance-state-name",
+                  "Values": ["running", "pending", "rebooting",
+                             "stopping", "stopped"]}])))
+    _LOGGER.debug("%d valid instances found", len(ec2_instances))
+
+    for hostname in ldap_servers - ec2_instances:
```

Comment:
nit - it is a bit difficult to read (i missed continue, and it took me a while to understand why are you deleting all servers.

how about a helper function (_valid_server(s): try gethostbyname(); return True; except gaierror: return False

and then filter?

also, i do not think that this function is correct. you are conflating two different independent sets - ldap and IPA.

for hostname in ldap_servers - ec2_instances:
   admin_srv.delete() - unconditional

for hostame in ipa_servers - ec2_instances:
 - delete.

i mean, if it is not in ec2 after one hour since defining in ldap - delete, regardless of dns, so i think dns check is not needed at all.

Reply:
refactored